util/logging.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package util
+
+import (
+	"flag"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/felixge/httpsnoop"
+	"github.com/gorilla/mux"
+	"go.elastic.co/ecszap"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Types of available loggers.
+const (
+	ECSLogger = "ecs"
+	DevLogger = "dev"
+
+	defaultLoggerType = ECSLogger
+)
+
+var logLevel = zap.LevelFlag("log-level", zap.InfoLevel, "log level (default \"info\")")
+var logType = flag.String("log-type", defaultLoggerType, "log type (ecs, dev)")
+
+var logger *zap.Logger
+var loggerMutex sync.Mutex
+
+// InitLogger initializes the logger, this is ignored afer a logger has
+// been created for this process.
+func InitLogger(loggerType string) {
+	loggerMutex.Lock()
+	defer loggerMutex.Unlock()
+
+	if logger != nil {
+		return
+	}
+
+	if loggerType == "" {
+		loggerType = *logType
+	}
+
+	switch loggerType {
+	case ECSLogger:
+		logger = newECSLogger()
+	case DevLogger:
+		logger = newDevelopmentLogger()
+	default:
+		logger = newECSLogger()
+		logger.Warn("unknown log type " + loggerType + " using default")
+	}
+}
+
+// Logger returns a logger singleton.
+func Logger() *zap.Logger {
+	InitLogger("")
+	return logger
+}
+
+func newECSLogger() *zap.Logger {
+	encoderConfig := ecszap.NewDefaultEncoderConfig()
+	core := ecszap.NewCore(encoderConfig, os.Stderr, *logLevel)
+	return zap.New(core, zap.AddCaller())
+}
+
+func newDevelopmentLogger() *zap.Logger {
+	encoderConfig := zap.NewDevelopmentEncoderConfig()
+	encoder := zapcore.NewConsoleEncoder(encoderConfig)
+	core := zapcore.NewCore(encoder, os.Stderr, zap.DebugLevel)
+	return zap.New(core, zap.AddCaller())
+}
+
+// LoggingMiddleware is a middleware used to log requests to the given logger.
+func LoggingMiddleware(logger *zap.Logger) mux.MiddlewareFunc {
+	// Disable logging of the file and number of the caller, because it will be the
+	// one of the helper.
+	logger = logger.Named("http").WithOptions(zap.WithCaller(false))
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Do not log requests to the health endpoint
+			if r.RequestURI == "/health" {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			LogRequest(logger, next, w, r)
+		})
+	}
+}
+
+// LogRequest captures information from a handler handling a request, and generates logs
+// using this information.
+func LogRequest(logger *zap.Logger, handler http.Handler, w http.ResponseWriter, req *http.Request) {
```

Comment:
nit: does it need to be exposed?

Reply:
Not really, changed. I considered to expose it because it is generic enough to be used with any handler and a zap logger. But no need to expose it by now.